streaming/storage/deque.go

```diff
@@ -0,0 +1,383 @@
+	"github.com/dgraph-io/badger/v2"
+	"github.com/golang/protobuf/proto"
+	"github.com/pkg/errors"
+)
+
+var ErrEmptyQueue = errors.New("the queue is empty")
+
+var (
+	dequeLastElementKey  = []byte("last")
+	dequeFirstElementKey = []byte("first")
+	dequeValueKeyPrefix  = []byte("value")
+	defaultValues        = map[string]int{
+		string(dequeFirstElementKey): 0,
+		string(dequeLastElementKey):  1,
+	}
+)
+
+type Deque struct {
+	tx           StateTransaction
+	initialized  bool
+	lastElement  int
+	firstElement int
+}
+
+type DequeIterator struct {
+	it Iterator
+}
+
+func NewDeque(tx StateTransaction) *Deque {
+	return &Deque{
+		tx:           tx,
+		initialized:  false,
+		lastElement:  1,
+		firstElement: 0,
+	}
+}
+
+func NewDequeIterator(it Iterator) *DequeIterator {
+	return &DequeIterator{
+		it: it,
+	}
+}
+
+//Adds new element to the front of the queue
+func (ll *Deque) PushFront(value proto.Message) error {
+	if !ll.initialized {
+		err := ll.initialize()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in queue.PushFront")
+		}
+	}
+
+	data, err := proto.Marshal(value)
+	if err != nil {
+		return errors.Wrap(err, "couldn't serialize given value")
+	}
+
+	return ll.appendBytes(data, ll.firstElement, ll.decFirst)
+}
+
+//Adds new element to the end of the queue
+func (ll *Deque) PushBack(value proto.Message) error {
+	if !ll.initialized {
+		err := ll.initialize()
+		if err != nil {
+			return errors.Wrap(err, "failed to initialize in queue.PushBack()")
+		}
+	}
+
+	data, err := proto.Marshal(value)
+	if err != nil {
+		return errors.Wrap(err, "couldn't serialize given value")
+	}
+
+	return ll.appendBytes(data, ll.lastElement, ll.incLast)
+}
+
+func (ll *Deque) appendBytes(data []byte, index int, modifier func() error) error {
+	byteKey := getIndexKey(index)
+
+	err := ll.tx.Set(byteKey, data)
+	if err != nil {
+		return errors.Wrap(err, "couldn't add the element to the queue")
+	}
+
+	err = modifier()
+	if err != nil {
+		return errors.Wrap(err, "failed to actualize queue attributes")
+	}
+
+	return nil
+}
+
+//Removes the first element from the queue and returns it
+//Returns ErrEmptyQueue if the list is empty
+func (ll *Deque) PopFront(value proto.Message) error {
+	//there is no need to call initialize here, since PeekFront calls initialize
+	err := ll.PeekFront(value)
+	if err != nil {
+		return err
+	}
+
+	return ll.pop(value, ll.firstElement+1, ll.incFirst)
+}
+
+//Removes the last element from the queue and returns it
+//Returns ErrEmptyQueue if the list is empty
+func (ll *Deque) PopBack(value proto.Message) error {
+	//there is no need to call initialize here, since PeekFront calls initialize
+	err := ll.PeekBack(value)
+	if err != nil {
+		return err
+	}
+
+	return ll.pop(value, ll.lastElement-1, ll.decLast)
+}
+
+func (ll *Deque) pop(value proto.Message, index int, modifier func() error) error {
+	err := ll.peekIndex(value, index)
+	if err != nil {
+		return err
+	}
+
+	key := getIndexKey(index)
```

Comment:
The idea of deleting firstElement + 1 is confusing for me -> it means, that firstElement is pointing at first free index instead of first element (same with last element).

Reply:
I think using
`return ll.pop(value, ll.lastElement, ll.decLast)`
instead of 
`return ll.pop(value, ll.lastElement-1, ll.decLast)`
would clarify this situation. This is the case in almost every function

Reply:
The problem is what happens when a queue is empty. At the beginning, an empty queue has it's first index at 0 and last index at 1, so it means the first "free" spots for PushFront and PushBack. Once you for example PushFront() then the first moves to -1, so to pop the first value, you actually have to add the +1. If we were to implement the first/last index as "position of first/last element" then it would become unclear what is to happen with an empty queue. I can try to reimplement it so it follows your suggestion and see how that goes

Reply:
Well, if that's problematic then change struct field names e.x. firstElement -> firstFrontFreeSpot or smth :x